src/shell/shell_misc.cpp

```diff
@@ -20,479 +20,421 @@
-					len = (uint16_t)it_history->length();
-					str_len = str_index = len;
-					size = CMD_MAXLINE - str_index - 2;
-					DOS_WriteFile(STDOUT, (uint8_t *)line, &len);
-					++it_history;
+			case Up:
+				if (history_index > 0) {
+					history_clone[history_index] = command;
+					history_index--;
+					command = history_clone[history_index];
+					cursor_position = command.size();
+				}
+				break;
 
+			case Down:
+				if (history_index < history_clone.size() - 1) {
+					history_clone[history_index] = command;
+					history_index++;
+					command = history_clone[history_index];
+					cursor_position = command.size();
+				}
+				break;
+
+			case Home: cursor_position = 0; break;
+
+			case End: cursor_position = command.size(); break;
+
+			case Delete: command.erase(cursor_position, 1); break;
+
+			case ShiftPlusTab:
+				if (completion.empty()) {
 					break;
-				case 0x53: /* Delete */
-					{
-						if(str_index>=str_len) break;
-						auto text_len = static_cast<uint16_t>(str_len - str_index - 1);
-						uint8_t* text=reinterpret_cast<uint8_t*>(&line[str_index+1]);
-						DOS_WriteFile(STDOUT, text, &text_len); // write buffer to screen
-						outc(' ');
-						move_cursor_back_one();
-						for (auto i = str_index; i < str_len-1; i++) {
-							line[i]=line[i+1];
-							move_cursor_back_one();
-						}
-						terminate_str_at(line, --str_len);
-						size++;
-					}
-					break;
-				case 15: /* Shift+Tab */
-					if (l_completion.size()) {
-						if (it_completion == l_completion.begin()) {
-							it_completion = l_completion.end ();
-						}
-						--it_completion;
-		
-						if (it_completion->length()) {
-							for (;str_index > completion_index; str_index--) {
-								// removes all characters
-								move_cursor_back_one();
-								outc(' ');
-								move_cursor_back_one();
-							}
-
-							strcpy(&line[completion_index], it_completion->c_str());
-							len = (uint16_t)it_completion->length();
-							str_len = str_index = completion_index + len;
-							size = CMD_MAXLINE - str_index - 2;
-							DOS_WriteFile(STDOUT, (uint8_t *)it_completion->c_str(), &len);
-						}
-					}
-				default: break;
 				}
-			};
-			break;
-		case 0x08: /* Backspace */
-			if (str_index) {
-				move_cursor_back_one();
-				size_t str_remain = str_len - str_index;
-				size++;
-				if (str_remain) {
-					memmove(&line[str_index-1],&line[str_index],str_remain);
-					terminate_str_at(line, --str_len);
-					str_index --;
-					/* Go back to redraw */
-					for (size_t i = str_index; i < str_len; i++)
-						outc(line[i]);
-				} else {
-					terminate_str_at(line, --str_index);
-					str_len--;
+				if (completion_index == 0) {
+					completion_index = completion.size();
 				}
-				outc(' ');
-				move_cursor_back_one();
-				// moves the cursor left
-				while (str_remain--)
-					move_cursor_back_one();
+				completion_index--;
+				command.erase(completion_start);
+				command += completion[completion_index];
```

Comment:
Are we guarantee that `completion_index` isn't already 0 when decrementing?  Also, is it guaranteed that the decremented completion_index is in-bounds of the `completion` vector?

Maybe some assert or use `.at(..)`?